### PR TITLE
Introduce domain setting to associate realms

### PR DESCRIPTION
**Pull request by albertzaharovits**
Introduces a new series of node settings `xpack.security.authc.domains.<domain_name>.realms: <realm_name_list>`.
The setting sets the `domain` property on `Realm` and `RealmConfig` instances.
The domain property ought be subsequently used in order to determine if two identical `usernames` are the same personas, and hence can share ownership (of profiles, keys, tokens, scrolls), even though they authenticated via different realms (but which are associated under the same domain).

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:350]: Why don't we use the list of `Realm`s built by `Realms#initRealms`? I think it could help simplify the logic, e.g. no specially handling for default file/native?
I guess one reason for this approach is that `domain` is a field under `RealmConfig` so it must be computed before realms are intialised. But if we can let go `domain` being field in `RealmConfig` and just have the whole domain concept works standalone, it will probably produce code that is easier to follow and requires less special logic.

**Review by ywangd — COMMENTED** [review]
I suggest we promote the implementation of security domain to be rather standalone instead being subordinate to realms. Concretely, I think we can:
1. Have a separate class for domain settings. I suspect there will be other domain setting down the line. One possible candidate that I can think of is a boolean setting to allow/disallow "phantom" realms. 
2. Instead of a static utility methods, add a new object (similar to `Realms`) that represents the domains. The object can help encapsulate validation logic. It has method that looks up domain information for a given `RealmRef`(a) and also looks up realm names for the given domain name. The object can then be injected into places where the lookup is needed. A full blown object provides better opportunities to make the interface more friendly to the callers (e.g. no need to know about settings and realmConfig) and more performant.
3. The above new "Domains" object can be instantiated after `Realms`. It takes information from `Realms`, but otherwise does not alter existing code of realms, including `RealmConfig`. If we allow "phantom" realms in future, there might not even be a `RealmConfig` for the domain information to attached to.

(a) It can also be `RealmIdentifier`. But thinking about the places where the method will be used, we are likely to have a `RealmRef` instead of `RealmIdentifier`. But I have no strong opinion on which one to use.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java:143]: After some more thoughts, I still think it's better if this method is not `static` but an instance method. We don't need a new class for it, it can just be a method of the existing Realms class. A few reasons:
* The comment says `Assumes {@code #verifyRealmNameToDomainNameAssociation} successfully verified the configuration.`. It means this method has a precondition that must be satisfied before invoking it. That's basically asking for some kinda of instantiation where object is properly prepared before answering questions.
* It simplies the method signature by dropping the `Settings` parameter.
* `Realms` knows about `List<Realm>` which is more efficient in finding out the realm domain. This method is needed by the Get Profile API, concretely, in `ProfileService#getProfile(String uid, ...)`. A more efficient lookup helps.
* For realms that are disabled, `Realms` is already computing a `List<RealmConfig>`. It is now thrown away after use. But we can easily keep it around to answer queries for disabled realms.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:350]: We discussed offline. The `List<Realm>` does not contain information about disabled realms. Hence the validation will not be complete with it.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java:136]: Why don't we apply this check to other synthetic realms?

**Review by ywangd — APPROVED** [review]
LGTM 

I'd appreciate if you could address the comments. I approved because the PR fulfills the requirements. Though I prefer methods like `getDomainForRealm` and future `getRealmsForDomain` to be instance methods instead of static methods, it does not have to block the PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java:48]: I think we can drop this validation entirely because:
* Realm names and deprecation are done elsewhere
* Domain and realm cross validation is also done elsewhere
* Realm name with leading underscore is not fully blocked so it is still a valid, though no-ideal, use case.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmsTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmsTests.java:357]: Made it a `TreeSet` 👍

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmsTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmsTests.java:275]: Okay, replaced with a List and used the default names for the two realms.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java:136]: Because `reserved` is a valid value for a realm name, that is not the reserved realm. In this case, the realm can be assigned to a domain but the legit reserved realm still must not be assigned to any realm.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java:48]: Okay, removed. I would think that because this is a new feature we could limit it to non-deprecated realms only, as a way to nudge realm renaming.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java]: I have pushed https://github.com/elastic/elasticsearch/pull/81968/commits/8a8a13cab84d4074503067557c4b4fe353a16db1 .

I still would've liked the realm to have a domain attribute, even though that might not be used in most places. It feels more OOP that way.

**Comment by albertzaharovits** [comment]
@ywangd I'm going to merge this unless you spot anything else.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java]: I have no issue for `Realm` to have a `domain` attribute. I actually think it's a good idea after our discussion. The comment was just about whether this method should be static and where it should be located. Sorry if that wasn't clear enough.

**Comment by ywangd** [comment]
> @ywangd I'm going to merge this unless you spot anything else.

Thanks for addressing all the feedback, Albert. I appreciate it. The only mis-communication is that the `domain` field got dropped from `Realm` which I didn't suggest. I think it's probably because `RealmConfig` can no longer call the static method anymore and things won't work straight given how `Realms` is now without some more refactoring. Maybe I was suggesting a bit more than what is really worth at this moment. I don't want to request further changes/refactoring for this PR. But given how it is now, I am happy for you to revert 8a8a13cab84d4074503067557c4b4fe353a16db1  Sorry for the back and forth.

**Comment by albertzaharovits** [comment]
Thanks Yang. Reverted and shipping.

**Pull request merged**
